### ISSUE TITLE
No Segmentation  

### ISSUE THREAD

**Issue by gorryfair** [ietf-tapswg/api-drafts#782]
Issue: No Segmentation  property - To me, the final text seems unclear about how it operates.:

   When set to true, this property requests the network layer at the
   sending endpoint to not fragment the packets generated by the
   transport layer.  When running over IPv4, setting this property to
   true will also cause the Don't Fragment bit to be set in the IP
   header.  When this property is set, an attempt to send a message size
   greater than the transport's current estimate of its maximum packet
   size (singularTransmissionMsgMaxLen) will result in a SendError.
   This only takes effect when the transport and network layer support
   this functionality.

- Is this really about the network fragmenting packets? - I do not think it is. To me, this is about the transport system segmenting packets.  I think this should read:

  When set to true, this property requests the transport system at the
   sending endpoint to not segment a single message into multiple IP packets.  
   When this property is set, an attempt to send a message
   greater in size than the transport's current estimate of its maximum packet
   size (singularTransmissionMsgMaxLen) will result in a SendError.
   When set to true, this also implies that noFragmentation is also set to true. 
   This only takes effect when the transport layer supports
   this functionality. Requiring this property prevents use of transports such as TCP.

**Comment by gorryfair** [comment]
Linked to #781

**Comment by mwelzl** [comment]
Good: I think you're right that this should not be about network-layer fragmentation.

Bad: The last sentence tells me that you wish for this to be a Selection Property, not (as it now is) a Message Property. From the minset appendix, we can see that the present property maps to "Request not to bundle messages" from SCTP; checking in RFC 8303, I find that this is the "no-bundle flag" of the send call, which we find in RFC 4960.  So that's what it is - a per-message request at the transport layer.

So, I suggest to update the text as follows (this only replaces the first two sentences with your new first sentence):

> When set to true, this property requests the transport system at the sending endpoint to not segment a single message into multiple IP packets. When this property is set, an attempt to send a message size greater than the transport's current estimate of its maximum packet size (singularTransmissionMsgMaxLen) will result in a SendError. This only takes effect when the transport and network layer support this functionality.

... and, if you NEED to be able to choose a transport that supports this functionality, look at the selection properties to see if you need to make a change to them. I'm not sure: we currently have "Preservation of Message Boundaries" there ("This property specifies whether the application needs or prefers to use a transport protocol that preserves message boundaries."), why is this not good enough?

**Comment by gorryfair** [comment]
OK. Similarly to the Frag text, I added the last sentence to try to tease-out what happens when you call the function. So, in this case what do what to happen for TCP?
* The message fails when sending with TCP - because it always segments?
* The message sends, because TCP always segments?
* Something else?

I'd really like a good example also of when the app would gain from setting this, possibly we could give an example of where setting this was useful? (e.g. an endpoint that wanted a single execution to process an entire packet?

**Comment by mwelzl** [comment]
I assume that you're talking about a message exceeding singularTransmissionMsgMaxLen. The current description is clear about this:
> When this property is set, an attempt to send a message size greater than the transport's current estimate of its maximum packet size (singularTransmissionMsgMaxLen) will result in a SendError.
I.e., a request from the application to send a message > singularTransmissionMsgMaxLen should fail, irrespective of the underlying transport. I say this just for context, to clarify that this is a discussion of the above sentence.

Now, let's think about whether this is a good design decision - should it REALLY fail in case of TCP?
I would say that the answer only is "yes" if guaranteeing "no fragmentation" is really useful for the application (which is the question you ask at the end).

Going back via the minset appendix and RFC 8303 allowed me to find the "no-bundle flag" of SCTP before, and now I started to dig for signs of its usefulness (assuming that we should be able to learn something from the design decisions that were made for SCTP). This made me understand that I made a mistake in the minset appendix, as I recently tried to adjust it to match updated names of Properties: this property does NOT fit the "no-bundle flag", because the "no-bundle flag" is essentially the same as enabling/disabling Nagle (with per-message control). We implement this via the capacity profile, it's a different thing.

Indeed, the "No Segmentation" functionality is not in minset, nor in RFC 8303 - neither SCTP nor TCP offer it. UDP does, implicitly, via the SEND call as described in RFC 8304. I can only imagine that an extremely time-critical application may want to execute this kind of control when knowing that it runs over UDP, but such an application could just make sure that their messages don't exceed the readable singularTransmissionMsgMaxLen property. Then that's a pretty UDP-specific implementation, I'd say, eliminating the need for a general "no segmentation" request that won't have a meaning anywhere else.

Am I missing something?  If everything I write above is correct, we should remove this property, UNLESS someone deems this super useful (i.e., just reading and adhering to singularTransmissionMsgMaxLen wouldn't be enough) when the underlying protocol is QUIC.
